src/components/select/QSelect.vue

```diff
@@ -195,68 +210,119 @@ export default {
-      return this.$refs.popover.hide()
-    },
-    reposition () {
+    onInputFilter () {
       const popover = this.$refs.popover
       if (popover.showing) {
         popover.reposition()
+        this.selectedIndex = this.$q.platform.is.desktop && this.enabledVisibleOptions.length !== 0
+          ? this.visibleOptions.findIndex(opt => !opt.disable)
+          : -1
+        this.$nextTick(this.scrollToSelectedItem)
       }
     },
-
-    __onFocus () {
-      this.focused = true
-      if (this.filter && this.autofocusFilter) {
-        this.$refs.filter.focus()
+    setCurrentSelection () {
+      if (this.selectedIndex >= 0 && !this.visibleOptions[this.selectedIndex].disable) {
+        this.multiple
+          ? this.__toggleMultiple(this.visibleOptions[this.selectedIndex].value)
+          : this.__singleSelect(this.visibleOptions[this.selectedIndex].value)
       }
-      this.$emit('focus')
-      const selected = this.$refs.popover.$el.querySelector(this.activeItemSelector)
+    },
+    scrollToSelectedItem (onShow = false) {
+      const selected = this.$refs.list.querySelector('.q-item.active')
       if (selected) {
-        selected.scrollIntoView()
+        let offset = 0
+        this.$refs.filter && (offset -= this.$refs.filter.$el.clientHeight)
+        onShow && (offset += this.$refs.list.clientHeight / 2)
+        const selectedTop = selected.offsetTop + offset
+        const selectedBottom = selected.offsetTop + selected.offsetHeight + offset
+        const listTop = this.$refs.list.scrollTop
+        const listBottom = listTop + this.$refs.list.clientHeight
+        if (selectedTop < listTop) {
+          this.$refs.list.scrollTop = selectedTop
+        }
+        else if (selectedBottom > listBottom) {
+          this.$refs.list.scrollTop = selectedBottom - this.$refs.list.clientHeight
+        }
       }
     },
-    __onBlur (e) {
-      this.__onClose()
-      setTimeout(() => {
-        const el = document.activeElement
-        if (el !== document.body && !this.$refs.popover.$el.contains(el)) {
-          this.hide()
-        }
-      }, 1)
+    move (offset) {
+      this.selectedIndex = normalizeToInterval(
+        this.selectedIndex + offset,
+        0,
+        this.visibleOptions.length - 1
+      )
+      this.visibleOptions[this.selectedIndex].disable === true ? this.move(offset) : this.$nextTick(this.scrollToSelectedItem)
     },
-    __onClose () {
-      this.focused = false
-      this.$emit('blur')
+    onShowPopover () {
+      this.selectedIndex = this.multiple
+        ? this.options.findIndex(opt => this.value.includes(opt.value))
+        : this.options.findIndex(opt => this.value === opt.value)
+      this.filter && this.$q.platform.is.desktop ? this.$refs.filter.focus() : this.$refs.list.focus()
+      this.$nextTick(this.scrollToSelectedItem.bind(null, true))
+      this.focused = true
+    },
+    onHidePopover () {
+      this.selectedIndex = -1
+      this.$refs.input.$el.focus()
```

Comment:
Still not sure about this... The whole idea is that the field remains in focus until the user leaves it explicitly.
So the field remains in focus even after popover hide by click on body or option. Simply put, hiding popover does not cause field lose focus. Of course, except if you click (tap) on another field or element which can take focus. I wonder if this is the correct behavior?
I would appreciate hearing other opinion on this ...

Reply:
Great work on this! I haven't looked at the code (I'm UX designer, FE development is new to me), looking forward to having this working though. 

How I would expect it to work is quite similar to the q-search with autocomplete. However, with one small difference and there might be something strange with the QAutocomplete? 

1. The QSelect gets focus: this triggers the popover to show (autocomplete only shows after searching, in this case the options are already known and can be shown right away after receiving focus)

2a. Selecting an option in the popover (through clicking or through arrowUp/arrowDown and Enter): popover hides and field loses focus. The @selected can be used to trigger any other actions (such as setting the focus on another item)

2b. The user clicks somewhere else (also if it is nothing) or uses tab: the QSelect loses focus. I would expect the popover to disappear. The autocomplete does that the same for clicking somewhere else. However, it doesnt for using tab to move to the next tabindexed item, in that case the popover stays open (until a search is started in another QSearch or the page scrolls). I don't know whether this is intended or a bug. I would expect (here and in the QAutocomplete): using tab moves to a next field and thus the QSelect loses focus and the popover is hidden.

Reply:
Thanks for your comment! We need to clarify something here:
1. When QSelect field gets focus by click - this triggers the popover to show.
But when QSelect field gets focus by press "Tab" on previous field then field only displayed as "focused" (Highlighted).
In this way user can "Tab" to the next field.
Press key "Enter" (or click of course) on such focused field  - triggers the popover to show.
Although the field remains highlighted but physically focus goes to the options list (or to the "filter" input on desktop platform, if it is here).
By the way @blur not emitted in this case like now.

2a & 2b
 - click on option (not multiple)
 - press "Enter" on option (not multiple)
 - press "Esc"
 - click somewhere else 

all of this triggers the popover to hide. Physically focus go to the field so user can "Tab" to the next field.  Because if the field lose focus the user will not be able navigate by "Tab".

And of course you can use any @events to trigger any other actions.

It is also possible to implement prop like "auto-open" for popover opened when field get focus.
But notice - in order to navigate by "Tab" popover need to be closed (by press "Esc" e.g.)
To make another behavior, need to move the search filter from popover to the field inline . 
But i'm not sure if this is correct ("a la material")

P.S. Of course QAutocomplete need to be fixed for hide popover when field lose focus.

Reply:
I have selected some parts from your comment:

**1. But when QSelect field gets focus by press "Tab" on previous field then field only displayed as "focused" (Highlighted). In this way user can "Tab" to the next field.**
Even if the popover would show, the user could still tab to the next field, am I correct? Then, I'd say the popover could also show automatically when focus is gained through Tab, this saves the user from an extra step (pressing the Enter). On the other hand: the popover could obscure the next item. So I could definitely also understand no auto-open. (implementing a prop would definitely be an option too, of course)

W.r.t. 2a & 2b: **all of this triggers the popover to hide. Physically focus go to the field so user can "Tab" to the next field. Because if the field lose focus the user will not be able navigate by "Tab".** Ah yes, it is indeed very important that the user can continue the tab from the place they were. However, I would say it depends on the interaction: if the user presses Esc: indeed: the field should stay active. If the users clicks randomly in the body: I would say then the field is not active anymore. The user might question if everything is working correctly if they click in nowhere and the focus is not lost. Also it might feel a bit 'trapped' to them: as if they have to input something there, even if they don't want to. 

**But notice - in order to navigate by "Tab" popover need to be closed (by press "Esc" e.g.)
To make another behavior, need to move the search filter from popover to the field inline. (like in QAutocomplete e.g.)
But i'm not sure if this is correct ("a la material")**
Hmm, this is important indeed. I would suggest that pressing tab should be enough to go to the next field: in that case it is similar to how the user works with the QAutocomplete. If a user presses Tab they would want to go to the next field. Maybe they wouldn't know about escape and thus continue with the mouse as soon as they meet something they would want to skip. I understand that you also have to take the extra time for moving the search filter to the field inline into account. I'm just giving my ideas on the subject :). 

I have tried searching for material guidelines for how to deal with tab-use, couldn't find much information (likely because on mobile, there is no tab). So I think it is left open for interpretation. 

What I'm certain of: auto-opening and needing escape to close would be bothersome. Especially with multiple optional Qselects it would cause irritation. 

Let's say a user has two optional fields, wants to set the first one and skip the next one:
A: Tab, (auto-open), arrows, Enter (to choose), Tab (now on the second field, auto-open), Tab (5 actions)
B: Tab, Enter (to open), arrows, Enter (to choose), Tab (now on the second field), Tab (6 actions)
Option A (auto-open, perhaps as prop, & Tab can be used to move to the next field, even with the popover open) saves the user an extra action. Also it prevents hindrance for users familiar with using Tab but who don't use Escape as often. 

PS: and using the @selected on the field to focus the next field would save another action, in both cases :D. I know, these are all details, but I like make things as easy as possible, even the small actions.

Reply:
>This was your original comment about, right, for which you were mostly looking for comments?

Aha, exactly ) But other comments are also very useful

>Even if the popover would show, the user could still tab to the next field, am I correct?

Neah... as i said earlier  - in order to navigate by "Tab" popover need to be closed. And it's natural. When popover is open, physically the focus is on it. But the field remains highlighted (not physically focused).

>I have tried searching for material guidelines for how to deal with tab-use

I meant a little bit different. The material concept is the layers. The popover layer is higher and you work with it, everything else under it should not be in focus. Popover casts a shadow on the bottom layer, indicating that he is now in charge. Therefore, the filter input is now placed inside it.

On the other hand, in order for everything to work as you wrote - need to move the filter inside the field inline like in QAutocomplete. By the way, this will also solve another problem. When you start typing something into the search, the height of the popover changes, and it repositions, jumps around the screen... This happens when the field is somewhere at the bottom of the screen, and the popover does not fit below, and is located at the top of the field.

I can suggest looking at the current version of this PR )

Reply:
Ah yes, your link makes it indeed very clear. Especially the select with filter is really nice, great addition! I'll be using it as soon as it becomes available. 

- something I noted is that on the colored versions: it is not very clear whether they are in focus or not. (With me: nothing changes at all). Or perhaps this is something still left to do. 

- The tab key currently gives some unexpected behaviour when used within anything with checkboxes or radio buttons: to reproduce: click on for example the multiple selection, press tab a few times, then Enter. It seems that the images of the checkbox/radio button are also part of the tabindex. However, this gives confusing behaviour to what is the currently selected item: the one with the arrows or the one with Tab.

- I found myself using Tab a few times with an open popover, to switch to the next field (also without thinking). However, the Tab then doesn't always go to a logical place (sometimes for instance to the secure connection icon in the addressbar). I think that using the tab to move to a next field would fit the users mental flow: I fill this in, go to the next one, skip this: I'll go to the next one. First having to close the popover would be okay looking at it from the code: you have opened something and first need to close it (as with brackets etc.). Most users are not that literal, I think they would think: I just want to go the next one and they'll expect the popover to close automatically when that field is not active anymore. I think to them, the field and the popover would be one element (in their mental model), not two separate entities. Especially on PC/laptop. On a smartphone, if the select menu would be a menu on top of everything else, and everything else faded out, then it is more clear that you first need to close it. In that case you wouldn't use a Tab key anyway. And for PC/laptop pages, the popover is not a fullscreen, everything else fades out, type of menu (nor do I think it should be). It is an element in a larger page. I hope I am clear with what I mean :). Just my 2 cents, trying ot help find what would be ideal. I think in any way there are very nice improvements!

Reply:
> select with filter is really nice, great addition

It is not my addition, actually this feature already implemented in current release e.g. (0.14.7)

> on the colored versions: it is not very clear whether they are in focus or not

Yes it is. This happens when "inverted" option is set on QSelect. This is native behavior and can be solved by custom css.

> It seems that the images of the checkbox/radio button are also part of the tabindex

Сheckbox/radio are focusable (have a tabindex). This can be fixed by adding to them props "tabindex" with value "-1". This should be done in the QSelect component itself. I'll try what can be done...

>However, the Tab then doesn't always go to a logical place

Of course) Because at this time the focus is on popover which physically located at the bottom of the document (at the end of DOM). I repeat myself: To make another behavior, need to move the search filter from popover to the field inline. (like in QAutocomplete e.g.) But i'm not sure that @rstoenescu will approve this.  As an option, may be to add a property such as "inline-filter" or somthing like this...

Reply:
>PPS: I searched, and for the QAutocomplete, there is already a bug report: #951

I just added PR #1238

Reply:
> select with filter is really nice, great addition
> It is not my addition, actually this feature already implemented in current release e.g. (0.14.7)

Ah yes, I noticed now, oops. Well, thanks for making me notice it then :)

> on the colored versions: it is not very clear whether they are in focus or not
> Yes it is. This happens when "inverted" option is set on QSelect. This is native behavior and can be solved by custom css.

Ah oké. I would suggest having something preset, perhaps to give them the same appearance as the QButtons. That way the out of the box functionality will also be clear. 

> It seems that the images of the checkbox/radio button are also part of the tabindex
Сheckbox/radio are focusable (have a tabindex). This can be fixed by adding to them props "tabindex" with value "-1". This should be done in the QSelect component itself. I'll try what can be done...
Great :). I wish I could help more with creating it also, at the moment I'm still learning to use Vue/Quasar, I think I would make a bit of a mess out of it. 

> However, the Tab then doesn't always go to a logical place
> Of course) Because at this time the focus is on popover which physically located at the bottom of the document (at the end of DOM). I repeat myself: To make another behavior, need to move the search filter from popover to the field inline. (like in QAutocomplete e.g.) But i'm not sure that @rstoenescu will approve this. As an option, may be to add a property such as "inline-filter" or somthing like this...
Ah now I understand better why that happens. Yes, perhaps @rstoenescu could add his ideas and opinions :)